controllers/provider_controller.go

```diff
@@ -44,61 +45,108 @@ import (
+	MaxConcurrentReconciles int
+}
+
+func (r *ProviderReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return r.SetupWithManagerAndOptions(mgr, ProviderReconcilerOptions{})
+}
+
+func (r *ProviderReconciler) SetupWithManagerAndOptions(mgr ctrl.Manager, opts ProviderReconcilerOptions) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1beta1.Provider{}).
+		WithEventFilter(predicate.Or(predicate.GenerationChangedPredicate{}, predicates.ReconcileRequestedPredicate{})).
+		WithOptions(controller.Options{MaxConcurrentReconciles: opts.MaxConcurrentReconciles}).
+		Complete(r)
 }
 
 // +kubebuilder:rbac:groups=notification.toolkit.fluxcd.io,resources=providers,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=notification.toolkit.fluxcd.io,resources=providers/status,verbs=get;update;patch
 
-func (r *ProviderReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	reconcileStart := time.Now()
-	log := logr.FromContext(ctx)
+func (r *ProviderReconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, retErr error) {
+	start := time.Now()
+	log := ctrl.LoggerFrom(ctx)
 
-	var provider v1beta1.Provider
-	if err := r.Get(ctx, req.NamespacedName, &provider); err != nil {
+	provider := &v1beta1.Provider{}
+	if err := r.Get(ctx, req.NamespacedName, provider); err != nil {
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
-	// record reconciliation duration
-	if r.MetricsRecorder != nil {
-		objRef, err := reference.GetReference(r.Scheme, &provider)
-		if err != nil {
-			return ctrl.Result{}, err
-		}
-		defer r.MetricsRecorder.RecordDuration(*objRef, reconcileStart)
+	r.RecordSuspend(ctx, provider, provider.Spec.Suspend)
+	// return early if the object is suspended
+	if provider.Spec.Suspend {
+		log.Info("Reconciliation is suspended for this object")
+		return ctrl.Result{}, nil
 	}
 
-	// validate provider spec and credentials
-	if err := r.validate(ctx, provider); err != nil {
-		provider = v1beta1.SetProviderReadiness(provider, metav1.ConditionFalse, meta.ReconciliationFailedReason, err.Error())
-		if err := r.patchStatus(ctx, req, provider.Status); err != nil {
-			return ctrl.Result{Requeue: true}, err
-		}
-		return ctrl.Result{Requeue: true}, err
+	patchHelper, err := patch.NewHelper(provider, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
 	}
 
-	if !apimeta.IsStatusConditionTrue(provider.Status.Conditions, meta.ReadyCondition) {
-		provider = v1beta1.SetProviderReadiness(provider, metav1.ConditionTrue, v1beta1.InitializedReason, v1beta1.InitializedReason)
-		if err := r.patchStatus(ctx, req, provider.Status); err != nil {
-			return ctrl.Result{Requeue: true}, err
+	defer func() {
+		patchOpts := []patch.Option{
+			patch.WithOwnedConditions{
+				Conditions: []string{
+					meta.ReadyCondition,
+					meta.ReconcilingCondition,
+					meta.StalledCondition,
+				},
+			},
+			patch.WithStatusObservedGeneration{},
```

Comment:
This is conditionally added on L108, and should be removed here to ensure the observed generation is only bumped once the resource has ended up in a "finite state" for that generation (by either successfully having reconciled to a `Ready==True` / `Stalled==True` /  `Reconciling==False` state).